### PR TITLE
Perf: load data systems on rank 0

### PR DESCRIPTION
**Pull request by caic99**
The current implementation loads data on each rank. This will stress the file system.
In this PR, only rank 0 will load data systems, and it will be broadcasted to each rank.
The data sampler initialized later will still use the exclusive seed of each rank.

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of distributed data loading for improved synchronization across processes.
	- Added broadcasting of the constructed dataset to ensure consistency in all processes.

- **Bug Fixes**
	- Implemented safeguards to prevent incomplete data distribution by asserting the integrity of the dataset.

### CODE REVIEW

File: deepmd/pt/utils/dataloader.py [hunk 1]
```diff
@@ -87,26 +87,24 @@ def __init__(
...
-                if dist.is_available() and dist.is_initialized()
-                else 1
-            )
-        ) as pool:
-            self.systems = pool.map(construct_dataset, systems)
-
+        self.systems: list[DeepmdDataSetForLoader] = []
+        global_rank = dist.get_rank() if dist.is_initialized() else 0
+        if global_rank == 0:
+            if len(systems) >= 100:
```
Comment (wanghan-iapcm) [inline deepmd/pt/utils/dataloader.py:99]: why do you need this condition?  maybe better to report the number of system no matter how many system you have

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Comment by njzjz** [comment]
> In this PR, only rank 0 will load data systems, and it will be broadcasted to each rank

Why do you expect each rank loads the same data?

**Comment by caic99** [comment]
> > In this PR, only rank 0 will load data systems, and it will be broadcasted to each rank
> 
> Why do you expect each rank loads the same data?

Good question. Currently every rank reads the config from the same `input.json` file, and initializes datasets there.

**Comment by njzjz** [comment]
> > > In this PR, only rank 0 will load data systems, and it will be broadcasted to each rank
> 
> > 
> 
> > Why do you expect each rank loads the same data?
> 
> 
> 
> Good question. Currently every rank reads the config from the same `input.json` file, and initializes datasets there.

I don't get your point. I believe the data is loaded asynchronously.

**Comment by caic99** [comment]
> > > Why do you expect each rank loads the same data?

> I don't get your point. I believe the data is loaded asynchronously.

I see. What you mean is not all ranks are accessing the same data files, and this is correct. The case is when the data is stored in a shared file system with a bandwidth limit.

**Comment by njzjz** [comment]
So it just affects the initialization of the data class? Well, I am not sure how much data is loaded in the initialization step.

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/pt/utils/dataloader.py [hunk 2]
```diff
@@ -87,26 +87,23 @@ def __init__(
...
-                else 1
-            )
-        ) as pool:
-            self.systems = pool.map(construct_dataset, systems)
-
+        self.systems: list[DeepmdDataSetForLoader] = []
+        global_rank = dist.get_rank() if dist.is_initialized() else 0
+        if global_rank == 0:
+            log.info(f"Constructing DataLoaders from {len(systems)} systems")
+            with Pool(os.cpu_count()) as pool:
```
Comment (njzjz) [inline deepmd/pt/utils/dataloader.py:100]: Why not using `env.NUM_WORKERS` here?

Comment (caic99) [reply deepmd/pt/utils/dataloader.py:100]:
```suggestion
            with Pool(env.NUM_WORKERS) as pool:
```

**Review by njzjz — APPROVED** [review]

File: deepmd/pt/utils/dataloader.py [hunk 3]
```diff
@@ -87,26 +87,23 @@ def __init__(
...
-                else 1
-            )
-        ) as pool:
-            self.systems = pool.map(construct_dataset, systems)
-
+        self.systems: list[DeepmdDataSetForLoader] = []
+        global_rank = dist.get_rank() if dist.is_initialized() else 0
+        if global_rank == 0:
+            log.info(f"Constructing DataLoaders from {len(systems)} systems")
+            with Pool(env.NUM_WORKERS) as pool:
```
Comment (njzjz) [inline deepmd/pt/utils/dataloader.py:100]:
```suggestion
            with Pool(max(1, env.NUM_WORKERS)) as pool:
```

**Pull request merged**